internal/pipeline/step_command_matrix.go

```diff
@@ -224,11 +226,38 @@ func (m *Matrix) validatePermutation(p MatrixPermutation) error {
 // being implicitly unique.
 type MatrixPermutation []SelectedDimension
 
+// UnmarshalJSON unmarshals a matrix permutation from a JSON object.
+func (mp *MatrixPermutation) UnmarshalJSON(b []byte) error {
+	if mp == nil {
+		return errors.New("unmarshal into nil pointer")
+	}
+	// Unmarshal the JSON using yaml.v3, to be consistent with everything else.
+	// For example, encoding/json prefers unmarshaling integer-looking numbers
+	// as float64.
+	m := make(map[string]any)
+	if err := yaml.Unmarshal(b, &m); err != nil {
+		return err
+	}
+	for dim, val := range m {
+		switch val.(type) {
+		case bool, int, string:
+			*mp = append(*mp, SelectedDimension{
+				Dimension: dim,
+				Value:     val,
+			})
+
+		default:
+			return fmt.Errorf("unsupported item type %T for key %q; want one of bool, int, or string", val, dim)
+		}
+	}
+	return nil
+}
```

Comment:
could we make the backend format matrix permutations in the way that we have them on the agent and avoid the extra unmarshalling logic here?

Reply:
That's reasonable. The backend already has to do similar work to massage the matrix back into the expected form. 

It wouldn't be enough to eliminate the method, because encoding/json will pick float64 for unmarshalling integers into any.

I was thinking, for simplicity, we could give up on preserving the type, and format ints and bools into strings as early as possible?

Reply:
i quite like that - i'm trying to think of a situation where munging stuff that's eventually going into a CLI into strings could be a problem, but i'm not coming up with much? that said, i'm sure there's something... 

it'd certainly be nice if the `Value` field of the `SelectedDimension` type was string.

Reply:
I think I'll merge this, then do a follow up.